File: tests/language/async_control_structures_test.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// VMOptions=--enable_async --optimization-counter-threshold=10

import 'package:expect/expect.dart';

import 'dart:async';

expectThenValue(future, value) {
  Expect.isTrue(future is Future);
  future.then((result) {
    Expect.equals(value, result);
  });
}

asyncIf(condition) async {
  if(condition) {
    return 1;
  } else {
    return 2;
  }
  // This return is never reached as the finally block returns from the
  // function.
  return 3;
}

asyncFor(condition) async {
  for (int i = 0; i < 10; i++) {
    if (i == 5 && condition) {
      return 1;
    }
  }
  return 2;
}

asyncTryCatchFinally(overrideInFinally, doThrow) async {
  try {
    if (doThrow) throw 444;
    return 1;
  } catch (e) {
    return e;
  } finally {
    if (overrideInFinally) return 3;
  }
}

asyncTryCatchLoop() async {
  var i = 0;
  var throws = 13;
  while (true) {
    try {
      throw throws;
    } catch (e) {
      if (i == throws) { return e; }
    } finally {
      i++;
    }
  }
}

asyncImplicitReturn() async {
  try {}
  catch (e) {}
  finally {}
}

main() {
  var asyncReturn;

  for (int i = 0; i < 10; i++) {
    asyncReturn = asyncIf(true);
    expectThenValue(asyncReturn, 1);
    asyncReturn = asyncIf(false);
    expectThenValue(asyncReturn, 2);

    asyncReturn = asyncFor(true);
    expectThenValue(asyncReturn, 1);
    asyncReturn = asyncFor(false);
    expectThenValue(asyncReturn, 2);

    asyncReturn = asyncTryCatchFinally(true, false);
    expectThenValue(asyncReturn, 3);
    asyncReturn = asyncTryCatchFinally(false, false);
    expectThenValue(asyncReturn, 1);
    asyncReturn = asyncTryCatchFinally(true, true);
    expectThenValue(asyncReturn, 3);
    asyncReturn = asyncTryCatchFinally(false, true);
    expectThenValue(asyncReturn, 444);
    asyncReturn = asyncTryCatchLoop();
    expectThenValue(asyncReturn, 13);

    asyncReturn = asyncImplicitReturn();
    expectThenValue(asyncReturn, null);
  }
}

File: pkg/dev_compiler/test/codegen_expected/language/async_control_structures_test.js
dart_library.library('language/async_control_structures_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__async_control_structures_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const async = dart_sdk.async;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const async_control_structures_test = Object.create(null);
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let dynamicAnddynamicTodynamic = () => (dynamicAnddynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic, dart.dynamic])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  async_control_structures_test.expectThenValue = function(future, value) {
    expect$.Expect.isTrue(async.Future.is(future));
    dart.dsend(future, 'then', dart.fn(result => {
      expect$.Expect.equals(value, result);
    }, dynamicTodynamic()));
  };
  dart.fn(async_control_structures_test.expectThenValue, dynamicAnddynamicTodynamic());
  async_control_structures_test.asyncIf = function(condition) {
    return dart.async(function*(condition) {
      if (dart.test(condition)) {
        return 1;
      } else {
        return 2;
      }
      return 3;
    }, dart.dynamic, condition);
  };
  dart.fn(async_control_structures_test.asyncIf, dynamicTodynamic());
  async_control_structures_test.asyncFor = function(condition) {
    return dart.async(function*(condition) {
      for (let i = 0; i < 10; i++) {
        if (i == 5 && dart.test(condition)) {
          return 1;
        }
      }
      return 2;
    }, dart.dynamic, condition);
  };
  dart.fn(async_control_structures_test.asyncFor, dynamicTodynamic());
  async_control_structures_test.asyncTryCatchFinally = function(overrideInFinally, doThrow) {
    return dart.async(function*(overrideInFinally, doThrow) {
      try {
        if (dart.test(doThrow)) dart.throw(444);
        return 1;
      } catch (e) {
        return e;
      }
 finally {
        if (dart.test(overrideInFinally)) return 3;
      }
    }, dart.dynamic, overrideInFinally, doThrow);
  };
  dart.fn(async_control_structures_test.asyncTryCatchFinally, dynamicAnddynamicTodynamic());
  async_control_structures_test.asyncTryCatchLoop = function() {
    return dart.async(function*() {
      let i = 0;
      let throws = 13;
      while (true) {
        try {
          dart.throw(throws);
        } catch (e) {
          if (i == throws) {
            return e;
          }
        }
 finally {
          i++;
        }
      }
    }, dart.dynamic);
  };
  dart.fn(async_control_structures_test.asyncTryCatchLoop, VoidTodynamic());
  async_control_structures_test.asyncImplicitReturn = function() {
    return dart.async(function*() {
      try {
      } catch (e) {
      }
 finally {
      }
    }, dart.dynamic);
  };
  dart.fn(async_control_structures_test.asyncImplicitReturn, VoidTodynamic());
  async_control_structures_test.main = function() {
    let asyncReturn = null;
    for (let i = 0; i < 10; i++) {
      asyncReturn = async_control_structures_test.asyncIf(true);
      async_control_structures_test.expectThenValue(asyncReturn, 1);
      asyncReturn = async_control_structures_test.asyncIf(false);
      async_control_structures_test.expectThenValue(asyncReturn, 2);
      asyncReturn = async_control_structures_test.asyncFor(true);
      async_control_structures_test.expectThenValue(asyncReturn, 1);
      asyncReturn = async_control_structures_test.asyncFor(false);
      async_control_structures_test.expectThenValue(asyncReturn, 2);
      asyncReturn = async_control_structures_test.asyncTryCatchFinally(true, false);
      async_control_structures_test.expectThenValue(asyncReturn, 3);
      asyncReturn = async_control_structures_test.asyncTryCatchFinally(false, false);
      async_control_structures_test.expectThenValue(asyncReturn, 1);
      asyncReturn = async_control_structures_test.asyncTryCatchFinally(true, true);
      async_control_structures_test.expectThenValue(asyncReturn, 3);
      asyncReturn = async_control_structures_test.asyncTryCatchFinally(false, true);
      async_control_structures_test.expectThenValue(asyncReturn, 444);
      asyncReturn = async_control_structures_test.asyncTryCatchLoop();
      async_control_structures_test.expectThenValue(asyncReturn, 13);
      asyncReturn = async_control_structures_test.asyncImplicitReturn();
      async_control_structures_test.expectThenValue(asyncReturn, null);
    }
  };
  dart.fn(async_control_structures_test.main, VoidTodynamic());
  // Exports:
  exports.async_control_structures_test = async_control_structures_test;
});
